Update ChaCha20 and ChaCha20Poly1305 cipher to match expected Cipher behaviour

KostasTsiounis opened the pull request:
Certain changes have to be made to match the official documentation regarding `Cipher` instances:
- In newer Java versions (11+), `init()` needs to be called after `doFinal()` when encrypting.
- When performing a `doFinal()` with a short buffer, a call to `init()` is required.

Tests to validate said behaviour are also added or altered accordingly.

Signed-off-by: Kostas Tsiounis <kostas.tsiounis@ibm.com>

taoliult commented on src/main/java/com/ibm/crypto/plus/provider/ChaCha20Cipher.java:
```diff
@@ -456,12 +366,9 @@ private void checkCipherInitialized() throws IllegalStateException {
-        this.initCalledInEncSeq = false;
+    // Reset class variables.
+    private void resetVars() {
+        this.initialized = (!this.encrypting); // force re-initialization only when decrypting
```
Should be `// force re-initialization only when encrypting`?

KostasTsiounis replied:
Yes. Fixed.

taoliult commented on src/main/java/com/ibm/crypto/plus/provider/ChaCha20Poly1305Cipher.java:
```diff
@@ -626,12 +506,10 @@ private void checkCipherInitialized() throws IllegalStateException {
-        this.initialized = false; // force re-initialization presumably with different nonce and key
+    // Reset class variables.
+    private void resetVars() {
+        this.initialized = (!this.encrypting); // force re-initialization only when decrypting
```
Also here, should be `// force re-initialization only when encrypting`?

KostasTsiounis replied:
Yes. Fixed.

jasonkatonica reviewed (commented):
Do we need an additional test update in `BaseTestChaCha20` to cover those cases given we changed both algorithms here?

KostasTsiounis commented:
> Do we need an additional test update in `BaseTestChaCha20` to cover those cases given we changed both algorithms here?

I added some further tests in `BaseTestChaCha20` to account for retries after a `ShortBufferException` occurs. See 40bd38bef1f8a964e18a0c26dc7394f6a0394fc0.
